### ISSUE TITLE
Map matching improvements

### ISSUE THREAD

**Issue by dabreegster** [nptscot/npw#40]
Map matching is used for traffic volumes, the core network, and the NPT flows. The CN results are currently poor. Red segments are detected CN segments, the black lines is the raw data.

![image](https://github.com/user-attachments/assets/b84f0a6d-df0d-4ae1-bffb-08c1c7f81192)
![image](https://github.com/user-attachments/assets/8cb80fd2-0038-489b-a2a7-2e773196c107)

I'm working on comparing candidate two linestrings by how parallel they are, the distance between endpoints, and the ratio of their lengths. Also considering Frechet and Hausdorff distance, but not seeing useful results from those yet.

**Comment by dabreegster** [comment]
![image](https://github.com/user-attachments/assets/4d1ca6fc-8717-4ab7-ac32-11e649c5680b)
Going to need more cleverness, because the OSM network is sometimes chopped into much smaller pieces than the CN. We could find the closest points on the longer line to the start/end of the smaller line, slice the longer line, and then do the angle/length comparisons.

Or buffer both and look at Jaccard index?

**Comment by dabreegster** [comment]
@Robinlovelace, do you have any advice from your other work on how to compare two linestrings for these cases? See above for things I've tried

**Comment by Robinlovelace** [comment]
Sure see here:

**Comment by Robinlovelace** [comment]
- Simple algorithm description: https://github.com/nptscot/rnetmatch/tree/main/rust
- Draft paper with more detail: https://github.com/nptscot/rnetmatch/blob/main/paper.qmd
- And Rust implementation: https://github.com/nptscot/rnetmatch/tree/main/rust/src

My sense is that this approach is good and works. It's not so much map matching as segment matching (it does not seek to correct GPS traces due to error and is a per-segment approach). What do you think?

**Comment by dabreegster** [comment]
Thanks for the pointers! I couldn't find a place in the npt codebase where you call this? What distance / angle thresholds have been useful for your cases so far?

Anyway, I tried this:

> We could find the closest points on the longer line to the start/end of the smaller line, slice the longer line, and then do the angle/length comparisons.

And it's incredibly effective:
![Screenshot from 2024-12-14 15-56-40](https://github.com/user-attachments/assets/056c8eb7-5149-49e3-ac85-ad7e394cd062)

There are some false positives now, which just look like distance/angle threshold tuning needed. I think my approach is similar in spirit to rnetmatch. I'll get things working well for the use cases here in the simplest way possible, then generalize the method / possibly integrate with rnetmatch and try to contribute some docs / code to it.

**Comment by dabreegster** [comment]
This is working super super well for CNs now. Some remaining exceptions:
Both sides of a dual carriageway by Water of Leith Centre match:
![image](https://github.com/user-attachments/assets/ed1980df-1bbc-471b-b9e2-dbc5c32c5706)
Some sharp angles break, like Blackhall path and Silverknowes:
![image](https://github.com/user-attachments/assets/7327d390-317a-4a90-a664-f9f897087341)
Roundabouts like Bullring:
![image](https://github.com/user-attachments/assets/ea2e4ecb-9fa0-4d9f-9bd8-01e31f07f1fb)

**Comment by Robinlovelace** [comment]
Fantastic, this looks great. Can you point us to the code that does this?

We us the following thresholds:

**Comment by Robinlovelace** [comment]
```r

  # Merge the spatial objects rnet_xp and rnet_yp based on specified parameters
  dist = 20
  angle = 15
  rnet_merged_all = stplanr::rnet_merge(rnet_xp, rnet_yp, dist = dist, funs = funs, max_angle_diff = angle, segment_length = 20)
```

Source: https://github.com/nptscot/npt/blob/00678ef7e551a5591cb809f315c84bb922a1e50a/R/simplify_network.R#L36C1-L40C129

**Comment by Robinlovelace** [comment]
That calls this function, see link to the source code: https://docs.ropensci.org/stplanr/reference/rnet_merge.html

Note: the source code there calls Rust code via the `rsgeo` R interface to the `geo` crate but does not use the `rnetmatch` crate currently (due to lack of time to cutover, not for lack of intention).

**Comment by Robinlovelace** [comment]
There is definitely a need to compare approaches and do some benchmarks, happy to take that on, OK for me to ping you @dabreegster for support implementing your approach?

Cc @josiahparry also, who will be interested in another approach implemented in Rust, great work getting it working from first principles.

**Comment by Robinlovelace** [comment]
Implementing or, ideally, using your approach : )

**Comment by dabreegster** [comment]
Thanks for the pointers! I cleaned up the code at https://github.com/nptscot/npw/blob/main/cli/src/match_lines.rs. The use for CNs is at https://github.com/nptscot/npw/blob/21c77eb5830b48acc676b997c45a17ed680292b5/cli/src/main.rs#L328; I'll use for traffic volumes and the rnet tomorrow, but I expect to see similarly good results, probably even using the same options.

The algorithm is quite simple and resembles my understanding of rnetmatch. All geometry is Euclidean. It's built for the CN case we've got, where almost always, the source is the same size or longer than the target:
![image](https://github.com/user-attachments/assets/a7f82187-d455-488a-8bfd-2e9ed319a0d1)
Might have to adjust things if that turns out false for the other use cases.

- For each target linestring (the road segments in the NPW graph):
- Calculate the AABB around the target, then buffer it by some amount (Otherwise, sources slightly away might be skipped entirely)
- Use an r-tree to find all source candidates in that AABB
- Slice the candidate to match up with the target better. Find the closest point on the source to the target's start and endpoint, then slice in between them
- Calculate the ratio of lengths between the candidates
- Calculate the difference in angle from the start -> endpoint (accounting for 180 degree differences of linestrings pointing opposite ways)
- Calculate the distance between midpoints
- Check these 3 metrics against user-specified thresholds
- If any of the candidates pass the thresholds, return a match, copying over user-provided data from the source.  (Right now it doesn't look for the _best_ match, just any match. Probably could combine the 3 metrics and take a minimum, if it's clear how to compare the tuple of 3 metrics or make a linear combo of them.)

There's a second variation of the method that can also dump one or many GeoJSON files for debugging all of this, which is how I wound up working through the problems developing this originally and how I tuned some of the thresholds so far. If this approach proves useful for the other things here, I have even more uses in other projects, and am thinking of a tiny little web app where you upload two GJ datasets, interactively tune the thresholds, and see results live. Let me know if you have any questions!

**Comment by JosiahParry** [comment]
@dabreegster that is _very_ close to the implementation. The difference is that rnetmatch builds an AABB around each line segment of the entire LineString. It’s intended for partial matching and provides a proportion of overlap. This allows you to choose how you perform the matching. 

The proposition of the draft paper in the repo is that it permits interpolation of values for partial matches. Classical map matching is focused on 1:1 matching which I don’t think is _always_ appropriate

**Comment by dabreegster** [comment]
Thanks for clarifying that. This makes sense -- in the use case here, one source maps to 1 or more targets. That's why I'm slicing the linestring to match the target. It's not important to know the percent of overlap, but I could imagine that elsewhere.

**Comment by Robinlovelace** [comment]
Definitely fuel for the implementation. Would be great if we could unify these things and get them in the `rnetmatch` crate, any blockers to using it atm Dustin? +1 to flexibility and generalisation, I imagine `rnetmatch` can do the all/nothing matching described, with the bonus that it can do partial matching which is awesome.

**Comment by Robinlovelace** [comment]
Hey @dabreegster we could actually use better map-matching on the R side, hence creating #45

**Comment by dabreegster** [comment]
![image](https://github.com/user-attachments/assets/55bc6a92-41b7-4b3f-9b90-64cc350ca46b)
Leith Walk in EDB is a good example of why different network representations at different stages gets incredibly confusing. The CN has just one simple line, but when interpreting in NPW, we have four choices -- two one-way roads, a cycletrack on both sides.

**Comment by Robinlovelace** [comment]
Indeed, Leith walk is a good example of when simplified representations are good for visualisation (but possibly not for routing).
